src/roster.ml

```diff
@@ -114,4 +115,45 @@ struct
     make_iq_request xmpp ?jid_from ?jid_to ?lang
       (IQGet (make_element (ns_roster, "query") [] []))
       callback
+
+  let put xmpp jid ?name ?groups ?jid_from ?jid_to ?lang ?(error_callback=ignore) callback =
+    let callback ev jid_from jid_to lang () =
+      match ev with
+      | IQResult el ->
+        callback ?jid_from ?jid_to ?lang
+      | IQError err ->
+        error_callback err
+    in
+    let attr = [make_attr "jid" jid]
+    in
+    let attr =
+      match name with
+      | Some x -> (make_attr "name" x) :: attr
+      | None   -> attr
+    in
+    let groupelem =
+      match groups with
+      | Some x -> List.map (fun e -> (make_element (no_ns, "group") [] [e])) x
+      | None   -> []
+    in
+    make_iq_request xmpp ?jid_from ?jid_to ?lang
+      (IQSet (make_element (ns_roster, "query") []
+                [make_element (no_ns, "item") attr groupelem]))
+      callback
+
+  let remove xmpp jid ?jid_from ?jid_to ?lang ?(error_callback=ignore) callback =
+    let callback ev jid_from jid_to lang () =
+      match ev with
+      | IQResult el ->
+        callback ?jid_from ?jid_to ?lang
+      | IQError err ->
+        error_callback err
+    in
+    let attr = [make_attr "jid" jid; make_attr "subscription" "remove"]
+    in
+    make_iq_request xmpp ?jid_from ?jid_to ?lang
+      (IQSet (make_element (ns_roster, "query") []
+                [make_element (no_ns, "item") attr []]))
+      callback
```

Comment:
this looks like a huge amount of code duplication... and why `put`, isn't it rather `set`?